lib/server/server.mjs

```diff
@@ -0,0 +1,406 @@
+* in the `context` object. Then, the next handler can perform operations on the
+* user, without needing to know how the user was authenticated.
+*
+* @callback RequestHandler
+* @param {Request} request Request data
+* @param {Response} response Response instance
+* @param {object} context Context object
+* @typedef {function(Request, Response, object)} RequestHandler
+*/
+
+/**
+* @summary Request exception handler callback.
+*
+* @description Exception handlers are called in cases when an exception is
+* thrown during request processing. Exception handlers act similar to regular
+* exception handlers, in the sense that a single request may go through multiple
+* exception handlers, and they may stream response data in return.
+*
+* They also have access to the same data as the regular request handlers, in
+* addition to the exception that occurred during processing.
+*
+* @callback RequestExceptionHandler
+* @param {Request} request Request data
+* @param {Response} response Response instance
+* @param {object} context Context object
+* @param {any} exception Exception occurred
+* @typedef {function(Request, Response, object, any)} RequestExceptionHandler
+*/
+
+/**
+* @summary Default request handler.
+*
+* @description This handler should be registered as a default handler, in cases
+* where no handlers are registered for the message subject. The response will
+* always be the following:
+*
+* ```js
+* {
+*   "header": {
+*     // original header
+*   },
+*
+*   "type": "err",
+*
+*   "error": {
+*     "type": "UnknownSubject",
+*     "message": "Unknown subject: <subject>"
+*   }
+* }
+* ```
+*
+* @param {Request} request Request data
+* @param {Response} response Response instance
+* @param {object} _context Context object
+*/
+export function unknownSubjectHandler (request, response, _context) {
+  response.error(new MessageError({
+    type: 'UnknownSubject',
+    message: `Unknown subject: ${request.header.subject}`
+  }))
+}
+
+/**
+* @summary Default exception handler.
+*
+* @description This exception handler is always registered and serves as a
+* catch-all exception handler. It will try to extract the exception's name and
+* message to be used as error type and message. Otherwise it will fall back to a
+* generic 'UnknownError' without message.
+*
+* @param {Request} _request Request data
+* @param {Response} response Response instance
+* @param {object} _context Context object
+* @param {any} exception Exception occurred
+*/
+export function defaultExceptionHandler (_request, response, _context, exception) {
+  if (exception.name && exception.message) {
+    response.error(exception.name, exception.message)
+  } else {
+    response.error('UnknownError', '')
+  }
+}
+
+/** @private */
+class StreamContext {
+  /** @type {string} */
+  id = nanoid()
+
+  /** @type {function(object)} */
+  handler
+
+  /** @type {stream.Duplex} */
+  stream
+
+  /** @type {stream.Duplex} */
```

Comment:
Just a pointless comment, but I love how JavaScript constantly shoots itself in the leg with "type hints"... :'D It's like, "we don't have strict typing, but it would be so good", then TypeScript winks back at JavaScript from far away. :'D

Reply:
I get your point, but in this case it's actually a good thing 😄 I'm only using Duplex as a 'placeholder' - as long as the thing being used can be piped and read like a stream, it will work.